### PR TITLE
Fix GRPO/DAPO/Dr.GRPO documentation: formula corrections and KL divergence clarification

### PR DESCRIPTION
**Pull request by JenWei0312**
This PR clarifies when to use KL divergence (β parameter) in GRPO/DAPO/Dr.GRPO implementations.

The current documentation has several issues:
1. It implies all loss types include KL divergence, contradicting the DAPO and Dr.GRPO papers
2. It has inconsistencies in mathematical formulas for GRPO
3. It lacks guidance on when to set β=0 vs β>0

Changes made:
- Corrected GRPO formula with proper normalization
- Clarified that DAPO and Dr.GRPO exclude KL term (β=0)
- Added a "To KL Or Not To KL" section explaining when KL can be omitted
- Improved formula presentation by separating KL from base loss term

References:
- DAPO: https://arxiv.org/abs/2503.14476 (Section 2.3)
- Dr.GRPO: https://huggingface.co/papers/2503.20783

# What does this PR do?

Fixes # (issue)

## Before submitting
- [X] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: docs/source/grpo_trainer.md [hunk 1]
```diff
@@ -106,7 +106,7 @@ Note that compared to the original formulation in [DeepSeekMath: Pushing the Lim
...
 
 $$
-\mathcal{L}_{\text{GRPO}}(\theta) = - \frac{1}{\sum_{i=1}^G |o_i|} \sum_{i=1}^G \sum_{t=1}^{|o_i|} \left[ \min \left( \frac{\pi_\theta(o_{i,t} \mid q, o_{i,< t})}{\pi_{\theta_{\text{old}}}(o_{i,t} \mid q, o_{i,< t})} \hat{A}_{i,t}, \, \text{clip}\left( \frac{\pi_\theta(o_{i,t} \mid q, o_{i,< t})}{\pi_{\theta_{\text{old}}}(o_{i,t} \mid q, o_{i,< t})}, 1 - \epsilon, 1 + \epsilon \right) \hat{A}_{i,t} \right) - \beta \mathbb{D}_{\text{KL}}\left[\pi_\theta \| \pi_{\text{ref}}\right] \right],
+\mathcal{L}_{\text{GRPO}}(\theta) =  - \frac{1}{G} \sum_{i=1}^G \frac{1}{|o_i|}  \sum_{t=1}^{|o_i|} \left[ \min \left( \frac{\pi_\theta(o_{i,t} \mid q, o_{i,< t})}{\pi_{\theta_{\text{old}}}(o_{i,t} \mid q, o_{i,< t})} \hat{A}_{i,t}, \, \text{clip}\left( \frac{\pi_\theta(o_{i,t} \mid q, o_{i,< t})}{\pi_{\theta_{\text{old}}}(o_{i,t} \mid q, o_{i,< t})}, 1 - \epsilon, 1 + \epsilon \right) \hat{A}_{i,t} \right) - \beta \mathbb{D}_{\text{KL}}\left[\pi_\theta \| \pi_{\text{ref}}\right] \right],
```
Comment (qgallouedec) [inline docs/source/grpo_trainer.md:109]: This would make the equation inconsistent with the previous one:
Screenshot 2025-05-01 at 10 44 49

File: docs/source/grpo_trainer.md [hunk 2]
```diff
@@ -117,13 +117,13 @@ When  \\( \mu = 1 \\) (default in TRL), the clipped surrogate objective simplifi
...
 
 $$
-\mathcal{L}_{\text{GRPO}}(\theta) = - \frac{1}{G} \sum_{i=1}^G \frac{1}{|o_i|} \sum_{t=1}^{|o_i|} l_{i,t},
+\mathcal{L}_{\text{GRPO}}(\theta) = - \frac{1}{G} \sum_{i=1}^G \frac{1}{|o_i|} \sum_{t=1}^{|o_i|} \left( l_{i,t} - \beta \mathbb{D}_{\text{KL}}\left[\pi_\theta \| \pi_{\text{ref}}\right] \right),
 $$
 
 where
 
 $$
-l_{i,t} = \frac{\pi_\theta(o_{i,t} \mid q, o_{i,< t})}{\left[\pi_\theta(o_{i,t} \mid q, o_{i,< t})\right]_{\text{no grad}}} \hat{A}_{i,t} - \beta \mathbb{D}_{\text{KL}}\left[\pi_\theta \| \pi_{\text{ref}}\right].
```
Comment (qgallouedec) [inline docs/source/grpo_trainer.md:126]: A bit the same here. I think it creates confusion, we mention the token-level normalisation, but when comparing the equations, the kl is also gone, without any explanation
 Screenshot 2025-05-01 at 10 47 48

File: docs/source/grpo_trainer.md [hunk 3]
```diff
@@ -141,6 +141,19 @@ $$
 
 This constant is recommended to be the maximum completion length. To use this formulation, set `loss_type="dr_grpo"` in the [`GRPOConfig`].
 
+#### To KL Or Not To KL
+Beside how we weight for each output loss by its length, both DAPO ([DAPO: An Open-Source LLM Reinforcement Learning System at Scale](https://arxiv.org/abs/2503.14476)) and Dr. GRPO ([Understanding R1-Zero-Like Training: A Critical Perspective](https://huggingface.co/papers/2503.20783)) exclude the KL term, or setting $\beta = 0$ in $\mathcal{L}_{\text{GRPO}}(\theta)$, because
+
+  1) during training the long-CoT reasoning model, the model distribution can diverge significantly from the initial model, thus this restriction is not necessary. And furthermore,
+  2) RL-tuning reasoning models typically employs rule-based verifiers as r [Tulu 3: Pushing Frontiers in Open Language Model Post-Training](https://arxiv.org/abs/2411.15124), eliminating the concerns of distributional shift.
```
Comment (qgallouedec) [inline docs/source/grpo_trainer.md:148]: I'm not sure how rule-based reward functions prevent distributional shift. Can you elaborate on this?

**Review by qgallouedec — COMMENTED** [review]
Thank you very much for improving the documentation, I still have some concerns about the clarity and fluidity of reading, I made some comments.

File: docs/source/grpo_trainer.md (hunk 3 above)
Comment (JenWei0312) [reply docs/source/grpo_trainer.md:148]: Regarding rule-based verifiers preventing distributional shift concerns: The key insight from the Dr. GRPO paper is that rule-based verifiers (like math answer checkers) provide accurate rewards regardless of how the policy distribution changes. Unlike learned reward models which are only accurate near their training distribution, rule-based verifiers are deterministic functions that don't suffer from distribution shift problems. This is why the papers argue it's safe to set β=0 when using rule-based rewards.

File: docs/source/grpo_trainer.md (hunk 2 above)
Comment (JenWei0312) [reply docs/source/grpo_trainer.md:126]: The removal of KL terms in the DAPO and Dr. GRPO formulations is explicitly stated in the original papers. In DAPO paper (section 2.3), they state: "during training the long-CoT reasoning model, the model distribution can diverge significantly from the initial model, thus this restriction is not necessary. Therefore, we will exclude the KL term from our proposed algorithm."

Similarly, the Dr. GRPO paper states on page 6 that we can "remove the KL term" when using rule-based verifiers. I wanted to make this distinction clear in the documentation since it represents an important theoretical difference between these algorithms.

I also attached both papers.
[DAPO.pdf](https://github.com/user-attachments/files/20028757/DAPO.pdf)
[Dr_GRPO.pdf](https://github.com/user-attachments/files/20028758/Dr_GRPO.pdf)

**Comment by JenWei0312** [comment]
Hi @qgallouedec 

I wanted to follow up on my previous responses about the KL divergence term and formula inconsistencies. I hope my explanations helped clarify my reasoning for the proposed changes.

After reflecting further on the documentation, I'm wondering if a more comprehensive restructuring might address these issues more effectively. Would you be open to an approach that clearly separates the theoretical background from the implementation details? Something like:

#### Research Papers Background
- GRPO (DeepSeekMath): Original formulation with sequence-level normalization and KL term
- DAPO: Token-level normalization without KL term
- Dr. GRPO: Constant normalization without KL term

#### Implementation Details
```python
if self.loss_type == "grpo":
    loss = ((per_token_loss * completion_mask).sum(-1) / completion_mask.sum(-1).clamp(min=1.0)).mean()
elif self.loss_type == "bnpo":
    loss = (per_token_loss * completion_mask).sum() / completion_mask.sum().clamp(min=1.0)
elif self.loss_type == "dr_grpo":
    loss = (per_token_loss * completion_mask).sum() / (per_token_loss.size(0) * self.max_completion_length)

File: docs/source/grpo_trainer.md (hunk 2 above)
Comment (ktolnos) [reply docs/source/grpo_trainer.md:126]: In the implementation, the presence of the KL is regulated only by beta=0, e.g. loss_type="dr_grpo" with beta=0.1 will still compute the KL divergence

**Comment by albertvillanova** [comment]
Thanks a lot for this @JenWei0312, and sorry for the very long silence on our side — this slipped through the backlog. 

Since you opened it, the GRPO docs were reworked and now cover exactly what you were after: there's a dedicated tip explaining that β=0 is the default and why the KL term is commonly dropped (citing DAPO, Dr.GRPO and Open-Reasoner-Zero), plus a Loss Types section covering DAPO/Dr.GRPO/SAPO. On top of that, the branch now conflicts, the normalization change to the clipped objective still has the formula-inconsistency you and @qgallouedec noted in review, and as @ktolnos pointed out the new tip isn't quite right — KL is gated by `beta=0`, not by `loss_type` (e.g. `dr_grpo` with `beta>0` still computes KL). 

So I'm going to close this one, but please don't take it as your work being wasted — if you spot a specific remaining inaccuracy in the current docs, a small focused PR against them would be very welcome!
